arrow-libs/core/arrow-core/src/commonTest/kotlin/arrow/core/continuations/EagerEffectSpec.kt

```diff
@@ -0,0 +1,164 @@
+    checkAll(Arb.int(), Arb.string()) { i, s ->
+      eagerEffect<String, Int> {
+        try {
+          shift(s)
+        } catch (e: Throwable) {
+          i
+        }
+      }.fold({ fail("Should never come here") }, ::identity) shouldBe i
+    }
+  }
+
+  "try/catch - finally works" {
+    checkAll(Arb.string(), Arb.int()) { s, i ->
+      val promise = CompletableDeferred<Int>()
+      eagerEffect<String, Int> {
+        try {
+          shift(s)
+        } finally {
+          require(promise.complete(i))
+        }
+      }
+        .fold(::identity) { fail("Should never come here") } shouldBe s
+      promise.await() shouldBe i
+    }
+  }
+
+  "try/catch - finally works suspended" {
+    checkAll(Arb.string(), Arb.int()) { s, i ->
+      val promise = CompletableDeferred<Int>()
+      eagerEffect<String, Int> {
+        try {
+          shift(s)
+        } finally {
+          require(promise.complete(i))
+        }
+      }
+        .fold(::identity) { fail("Should never come here") } shouldBe s
+      promise.await() shouldBe i
+    }
+  }
+
+  "try/catch - First shift is ignored and second is returned" {
+    checkAll(Arb.int(), Arb.string(), Arb.string()) { i, s, s2 ->
+      eagerEffect<String, Int> {
+        val x: Int =
+          try {
+            shift(s)
+          } catch (e: Throwable) {
+            i
+          }
+        shift(s2)
+      }
+        .fold(::identity) { fail("Should never come here") } shouldBe s2
+    }
+  }
+
+  suspend fun test() {
+    val leakedAsync =
+      coroutineScope<suspend () -> Deferred<Unit>> {
+        suspend {
+          async { println("I am never going to run, until I get called invoked from outside") }
+        }
+      }
+    leakedAsync.invoke().await()
+  }
+
+  "try/catch - First shift is ignored and second is returned suspended" {
+    checkAll(Arb.int(), Arb.string(), Arb.string()) { i, s, s2 ->
+      eagerEffect<String, Int> {
+        val x: Int =
+          try {
+            shift(s)
+          } catch (e: Throwable) {
+            i
+          }
+        shift(s2)
+      }
+        .fold(::identity) { fail("Should never come here") } shouldBe s2
+    }
+  }
+
+  "immediate values" { eagerEffect<Nothing, Int> { 1 }.value() shouldBe 1 }
+
+  "suspended value" { eagerEffect<Nothing, Int> { 1 }.value() shouldBe 1 }
+
+  "immediate short-circuit" {
+    eagerEffect<String, Nothing> { shift("hello") }.runCont() shouldBe "hello"
+  }
+
+  "suspended short-circuit" {
+    eagerEffect<String, Nothing> { shift("hello") }.runCont() shouldBe "hello"
+  }
+
+  "Rethrows immediate exceptions" {
+    val e = RuntimeException("test")
+    Either.catch { eagerEffect<Nothing, Nothing> { throw e }.runCont() } shouldBe Either.Left(e)
+  }
+
+  "Rethrows suspended exceptions" {
+    val e = RuntimeException("test")
+    Either.catch { eagerEffect<Nothing, Nothing> { throw e }.runCont() } shouldBe Either.Left(e)
+  }
+
+  // Fails https://github.com/Kotlin/kotlinx.coroutines/issues/3005
```

Comment:
Can we fix this by not using the method reference for `inline suspend` somewhere?

Reply:
yeah this was not failing